### ISSUE TITLE
`minigallery` with multiple files: deduplicate and should not start new rows

### ISSUE THREAD

**Issue by Charlie-XIAO** [sphinx-gallery/sphinx-gallery#1259]
When using `minigallery` with multiple files, each file will start a new row. It would be nice if they do not start new rows. Also, it would be nice to remove duplicates if multiple files share a same example. The old behavior can be easily retained by calling `minigallery` for the files separately.

![image](https://github.com/sphinx-gallery/sphinx-gallery/assets/108576690/e59e693f-1dca-432b-98d2-e77306433a06)

**Comment by lucyleeow** [comment]
@story645 I thought we fixed the separate lines issue? 

The duplicates is known. I don't know if there is an easy fix because the duplicates could come from backreference, so we'd have to dig into the `.example` file itself, and then we may have to amend  the contents of one if e.g., 2 `.example` files both include an example  @larsoner any ideas?

**Comment by lucyleeow** [comment]
Yes good point, sorry I saw 'file' and thought of the new directive but you're right, we haven't released yet. I guess it's time for a new release, WDYT @larsoner ?

> One of the things I've been thinking about is if there's any appetite for maintaining the back references in something like a table or graph?

That sounds reasonable to me, just a lot of work.

What do you mean by double parsing? Thanks

**Comment by lucyleeow** [comment]
Oh yes I remember that. It would be good to store the title with the backreference. Agree, backreferences could definitely be improved and table/graph sound like good options.

**Comment by lucyleeow** [comment]
> maintaining the back references in something like a table or graph?

@larsoner , this seems like a reasonable way to fix the duplicate example problem with the minigallery directive.

If I store the backreferences in a table/dict instead, should/can I remove the '.examples' backreferences file, or do you think some users rely on these files and thus we should keep it to remain backwards compatible?

**Comment by larsoner** [comment]
> should/can I remove the '.examples' backreferences file, or do you think some users rely on these files and thus we should keep it to remain backwards compatible?

A few years ago we recommended to embed the `.examples` files using an `.. include` directive, so it's possible people still use them. For example here's the PR for MNE 4.5 years ago where we went from the "old way" to the "new way"

https://github.com/mne-tools/mne-python/commit/5701738ba10d65e2c5a713e0c6480aa04cd1a4ae#diff-139831090703b930856dfa049e8bb35d602bf5822d01f4d344efc2c9d4f94578

No idea how many people still use the "old way" but if it's easy enough to keep them around we might as well

**Comment by lucyleeow** [comment]
Good to know! I an attempt at searching for ".examples" in github was not fruitful.

Looking into this more, it seems that internally in SG we use these example files in 2 places:

* minigallery directive - to fix the duplication problem we would want parse to get list of examples
* API usage - info required is whether object is unused, and if used, list of example files

Indeed for API usage we already read in the .examples file and parse it to get the list of examples:

https://github.com/sphinx-gallery/sphinx-gallery/blob/6c4849583c120b76843e7dcffc749e9e56cada2a/sphinx_gallery/gen_gallery.py#L1242-L1246

Since both use cases inside SG now need the list of examples, and would need to parse the .examples file to get this, it would probably (?) be more efficient to store the the backreferences as a e.g., dict. But this would not be backwards compatible.

It seems our options are:

* keep .examples file, parse it each time we use it
* change to storing backreferences as dict and keep producing the '.examples' files for backwards compat
* change to storing backreferences as dict but stop producing the '.examples' files

WDYT ?

**Comment by larsoner** [comment]
> change to storing backreferences as dict and keep producing the '.examples' files for backwards compat

I think this one is the cleanest solution. Keep things in `python` but keep writing the files (probably at the end of the SG run or whenever deduplication is complete) because it's easy enough to do

**Comment by lucyleeow** [comment]
> unless parallelization is problematic for some reason

AFAICT we parallelize running `generate_file_rst` (create rst for single example), but we write backreferences outside of parallel, in `generate_dir_rst` (create rst for single folder i.e. root gallery or sub-gallery).

I think (?) this means that if I were to create a single dict, with all backreferences, inside `generate_gallery_rst`, it would not be a problem for parallelization? 

> depending on how large the dicts gets,

Roughly, how large would this be?
